packages/relay/src/lib/eth.ts

```diff
@@ -1320,24 +1320,31 @@ export class EthImpl implements Eth {
       return [];
     }
 
-    return result.logs.map(log => {
-      return new Log({
-        address: log.address,
-        blockHash: EthImpl.toHash32(log.block_hash),
-        blockNumber: EthImpl.numberTo0x(log.block_number),
-        data: log.data,
-        logIndex: EthImpl.numberTo0x(log.index),
-        removed: false,
-        topics: log.topics,
-        transactionHash: EthImpl.toHash32(log.transaction_hash),
-        transactionIndex: EthImpl.numberTo0x(log.transaction_index)
-      });
-    });
+    return Promise.all(
+      result.logs.map(async log => {
+        return new Log({
+          address: await this.getContractEvmAddress(log.address, requestId),
+          blockHash: EthImpl.toHash32(log.block_hash),
+          blockNumber: EthImpl.numberTo0x(log.block_number),
+          data: log.data,
+          logIndex: EthImpl.numberTo0x(log.index),
+          removed: false,
+          topics: log.topics,
+          transactionHash: EthImpl.toHash32(log.transaction_hash),
+          transactionIndex: EthImpl.numberTo0x(log.transaction_index)
+        });
+      })
+    );
   }
 
   async maxPriorityFeePerGas(requestId?: string): Promise<string> {
     const requestIdPrefix = formatRequestIdMessage(requestId);
     this.logger.trace(`${requestIdPrefix} maxPriorityFeePerGas()`);
     return EthImpl.zeroHex;
   }
+
+  private async getContractEvmAddress(address: string, requestId: string | undefined) {
+    const result = await this.mirrorNodeClient.getContract(address, requestId);
```

Comment:
Can we cache this lookup with some fixed size LRU cache?

Reply:
Can use something like
- check for `contractEvmAddress` i.e 
```
 let contractEvmAddress = cache.get(cachedLabel);
```
- if absent make mirror node call and when done.
```
cache.set(cachedLabel, <contractEvmAddress>, constants.CACHE_TTL.ONE_HOUR);
```

Reply:
I've added the caching, added a test. 

One question, do we need the `result.evm_address || address`, it seems `evm_address` can't be `null` unless `result` is null, in which case `result.evm_address` will throw a `TypeError: Cannot read properties of null (reading 'evm_address')`. Also in this case `result` would be `null` if the mirror node returns 400 or 404, wouldn't we want  to throw an error in the case when the request result is `null`?

Reply:
I think it might be better to catch any errors during `getContractEvmAddress()` and return the `log.address` as a fallback.